Feature/PanelDue

AHHams opened the pull request:
This implements a reporter feature to provide the necessary JSON responses to the M408 code in order to interoperate with PanelDue. 

It aims to implement the JSON spec as defined here: [Old JSON response spec](http://reprap.org/wiki/G-code#M408:_Report_JSON-style_response). It does not yet support the [new spec](http://reprap.org/wiki/RepRap_Firmware_Status_responses). But that would be easy to add as second step. 

The code is based on the original pull request by Chris Brent, that was closed. This version has moved on from there quite bit and supports more of the JSON spec. 

The code works and communication with PanelDue works well (other than for points mentioned below). Latest version of PanelDue firmware (1.15c) is needed. I have done limited testing to date.
 
There are a few issues / missing features (marked with "TODO" in comments in code). However, since the basics work, I decided to trigger a pull request. Also in order to solicit some feedback, and pointers as to how I might implement missing features.

Code still has following issues:
1. PanelDue set extruder temp using G10 rather then M104. Solving this, in my view, also requires change at PanelDue side, so I initiated a discussion.
2. Does not correctly deal with heated chambers (neither does PanelDue, I think), so not urgent.
3. I can only get the fan percentage to return either 0% or 100%. 
4. responses to "status" could probably be made more complete.
5. Response to seq/resp just a stub. How to I get latest non-trivial response? (needed to support PanelDue's console).
6. "standby" temperature not supported as concept is foreign to Smoothie (AFAIK).
7.  I don't know how to respond to "extr" ("extruder total extrusion since power up, last G92 or last M23 ") 
8. I don't know how to respond to "efactor" (extruder override factor).
9. I don't know how to respond to "zprobe", "fanRPM".

Please let me know what you think.

Commit:
Added Reporter class for m408 gcode (e.g. DuePanel)

chrisbrent committed:
Merge pull request #1 from Smoothieware/edge

Update to current Edge

chrisbrent committed:
Merge pull request #2 from Smoothieware/edge

Merging current edge

Commit:
Setup TemperatureControlPool for access by the Reporter class

Commit:
Merged branch feature/due_panel into feature/due_panel

Commit:
Reverting unneeded changes. Changed Reporter Module to get temperatures properly.

Commit:
Blank line removed

Commit:
Add heater status.

Commit:
Add config to load Reporter module. Add status, see code for assumption of M408 S0 status mapping to Smoothie data.

Commit:
Fix complilation error

Commit:
Clean up Json so that it's valid. Change vector auto iterators to index based.

Commit:
Fix for heater status.

chrisbrent committed:
Merge pull request #3 from Smoothieware/edge

Merge edge

chrisbrent committed:
Merge pull request #4 from chrisbrent/feature/due_panel

Merge edge into Feature/due panel

chrisbrent committed:
Merge pull request #5 from chrisbrent/edge

Latest edge to due_panel

Commit:
Merged branch feature/due_panel into feature/due_panel

Commit:
Add zprobe height.

Commit:
Don't read Zprobe height from the config file, because Smoothie crashes.... Need to hook up a debugger to see why

chrisbrent committed:
Merge pull request #6 from Smoothieware/edge

Updating edge with Smoothieware

chrisbrent committed:
Merge pull request #7 from Smoothieware/edge

Merge edge into Feature/due panel

Commit:
Merge edge

Commit:
Merge branch 'Smoothieware-edge'

chrisbrent committed:
Merge pull request #9 from Smoothieware/edge

Pull lastest edge.

chrisbrent committed:
Merge pull request #10 from Smoothieware/edge

Getting Edge back in sync with my branch

Commit:
Sort the heaters so that the bed is first.

Commit:
Add changes to main.cpp to allow reporter module to load.

chrisbrent committed:
Merge pull request #11 from Smoothieware/edge

Merge edge

Commit:
Merged branch Smoothieware-edge into feature/due_panel

chrisbrent committed:
Merge pull request #12 from Smoothieware/edge

Update to latest edge

chrisbrent committed:
Merge branch 'edge'

Conflicts:
	src/main.cpp

AHHams committed:
Fixed merge conflict

AHHams committed:
Fix conflict with respect to Endstops() line

AHHams committed:
Rewrite, based on Chris Brent's version + inspiration from David Crocker's code.

AHHams committed:
Merge remote-tracking branch 'upstream/edge' into edge

AHHams committed:
Merge remote-tracking branch 'upstream/edge' into feature/due_panel

AHHams committed:
Merge branch 'edge' into feature/due_panel

AHHams committed:
Implemented a stub for seq and resp. Some minor changes.

AHHams committed:
Added an example snippet showing how to configure the Reporter module.

arthurwolf commented:
Is this compiled by default ? I think we probably want to have an option
not to compile it in by default. the current firmware is pretty tight on
flash and not many people have a paneldue.

On Tue, Nov 8, 2016 at 11:23 PM, AHHams <notifications@github.com> wrote:

> This implements a reporter feature to provide the necessary JSON responses
> to the M408 code in order to interoperate with PanelDue.
>
> It aims to implement the JSON spec as defined here: Old JSON response spec
> <http://reprap.org/wiki/G-code#M408:_Report_JSON-style_response>. It does
> not yet support the new spec
> <http://reprap.org/wiki/RepRap_Firmware_Status_responses>. But that would
> be easy to add as second step.
>
> The code is based on the original pull request by Chris Brent, that was
> closed. This version has moved on from there quite bit and supports more of
> the JSON spec.
>
> The code works and communication with PanelDue works well (other than for
> points mentioned below). Latest version of PanelDue firmware (1.15c) is
> needed. I have done limited testing to date.
>
> There are a few issues / missing features (marked with "TODO" in comments
> in code). However, since the basics work, I decided to trigger a pull
> request. Also in order to solicit some feedback, and pointers as to how I
> might implement missing features.
>
> Code still has following issues:
>
>    1. PanelDue set extruder temp using G10 rather then M104. Solving
>    this, in my view, also requires change at PanelDue side, so I initiated a
>    discussion.
>    2. Does not correctly deal with heated chambers (neither does
>    PanelDue, I think), so not urgent.
>    3. I can only get the fan percentage to return either 0% or 100%.
>    4. responses to "status" could probably be made more complete.
>    5. Response to seq/resp just a stub. How to I get latest non-trivial
>    response? (needed to support PanelDue's console).
>    6. "standby" temperature not supported as concept is foreign to
>    Smoothie (AFAIK).
>    7. I don't know how to respond to "extr" ("extruder total extrusion
>    since power up, last G92 or last M23 ")
>    8. I don't know how to respond to "efactor" (extruder override factor).
>    9. I don't know how to respond to "zprobe", "fanRPM".
>
> Please let me know what you think.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/Smoothieware/Smoothieware/pull/1059
> Commit Summary
>
>    - Added Reporter class for m408 gcode (e.g. DuePanel)
>    - Merge pull request #1 from Smoothieware/edge
>    - Merge pull request #2 from Smoothieware/edge
>    - Setup TemperatureControlPool for access by the Reporter class
>    - Merged branch feature/due_panel into feature/due_panel
>    - Reverting unneeded changes. Changed Reporter Module to get
>    temperatures properly.
>    - Blank line removed
>    - Add heater status.
>    - Add config to load Reporter module. Add status, see code for
>    assumption of M408 S0 status mapping to Smoothie data.
>    - Fix complilation error
>    - Clean up Json so that it's valid. Change vector auto iterators to
>    index based.
>    - Fix for heater status.
>    - Merge pull request #3 from Smoothieware/edge
>    - Merge pull request #4 from chrisbrent/feature/due_panel
>    - Merge pull request #5 from chrisbrent/edge
>    - Merged branch feature/due_panel into feature/due_panel
>    - Add zprobe height.
>    - Don't read Zprobe height from the config file, because Smoothie
>    crashes.... Need to hook up a debugger to see why
>    - Merge pull request #6 from Smoothieware/edge
>    - Merge pull request #7 from Smoothieware/edge
>    - Merge edge
>    - Merge branch 'Smoothieware-edge'
>    - Merge pull request #9 from Smoothieware/edge
>    - Merge pull request #10 from Smoothieware/edge
>    - Sort the heaters so that the bed is first.
>    - Add changes to main.cpp to allow reporter module to load.
>    - Merge pull request #11 from Smoothieware/edge
>    - Merged branch Smoothieware-edge into feature/due_panel
>    - Merge pull request #12 from Smoothieware/edge
>    - Merge branch 'edge'
>    - Fixed merge conflict
>    - Fix conflict with respect to Endstops() line
>    - Rewrite, based on Chris Brent's version + inspiration from David
>    Crocker's code.
>    - Merge remote-tracking branch 'upstream/edge' into edge
>    - Merge remote-tracking branch 'upstream/edge' into feature/due_panel
>    - Merge branch 'edge' into feature/due_panel
>    - Implemented a stub for seq and resp. Some minor changes.
>    - Added an example snippet showing how to configure the Reporter
>    module.
>
> File Changes
>
>    - *A* ConfigSamples/Snippets/reporter.config
>    <https://github.com/Smoothieware/Smoothieware/pull/1059/files#diff-0>
>    (2)
>    - *M* src/main.cpp
>    <https://github.com/Smoothieware/Smoothieware/pull/1059/files#diff-1>
>    (4)
>    - *A* src/modules/tools/reporter/Reporter.cpp
>    <https://github.com/Smoothieware/Smoothieware/pull/1059/files#diff-2>
>    (374)
>    - *A* src/modules/tools/reporter/Reporter.h
>    <https://github.com/Smoothieware/Smoothieware/pull/1059/files#diff-3>
>    (24)
>
> Patch Links:
>
>    - https://github.com/Smoothieware/Smoothieware/pull/1059.patch
>    - https://github.com/Smoothieware/Smoothieware/pull/1059.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Smoothieware/Smoothieware/pull/1059>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAGpFdCjMSkH1fwRj-n_w9YG6QTDbDPpks5q8PZ-gaJpZM4Ks-zq>
> .
>



-- 
Courage et bonne humeur.

wolfmanjm commented:
The problem I see is that panedue is designed to work with a totally different firmware that differs significantly from smoothie, and I am not in favor of adding concepts to smoothie that it would need.

G10 is a deal killer for me that is not an option to change in smoothie. it is already overloaded as both firmware retract and setting WCS.  that needs to be changed on paneldue.

AHHams commented:
@arthurwolf: I am new to firmware programming. If there's anything i can do to minimise memory footprint of the compiled code, let me know (the repeated printf calls can be compacted into a single one, at the expense of code readability...). The limited nr. of paneldue owners might have something to do with Smoothie lacking support. :-) Moreover, the M408 / JSON output might also serve webinterfaces, once feature complete. 

@wolfmanjm: I don't think PanelDue is _very_ tightly coupled to RepRapFirmware specifics. Though in some places indeed it is. Notably wrt G10, and "standby" temperatures. It may also be in David's interest to make / allow improvements to the PanelDue code to make it more generic. As removing the tight coupling to RepRapFirmware might increase the potential user base quite bit.

AHHams committed:
Add firmwareName to JSON output, so PanelDue can make behavior specific to Smoothie.

AHHams committed:
Merge remote-tracking branch 'Smoothieware/edge' into feature/due_panel

wolfmanjm commented:
- _I don't know how to respond to "extr" ("extruder total extrusion since power up, last G92 or last M23 ")_
  G92 is used to set global workspace coordinate system, it can be queried with `get wcs` but I doubt that is what paneldue thinks it is, it probably uses G92 for some other thing
  M23 can be found just look at the watchscreen.cpp code to see how to get currently playing file

- _I don't know how to respond to "efactor" (extruder override factor)._
   panel does this so look at how the ExtruderScreen gets the E scaling factor

- _I don't know how to respond to "zprobe", "fanRPM"._
   there is a way to get a switches PWM value using `PublicData::get_value( switch_checksum, get_checksum(type), value_checksum, &v );` which would be proportional to fan RPM, However there is no way to know what switch is a fan, so basically this in not really very easy to do,

  I doubt any of the zprobe stuff is compatible with smoothie, however there is a panel Zprobe menu you can look at that to see what info can be extracted.

wolfmanjm commented:
I presume the Paneldue connects to the serial uart? If so how does it handle the data we spit out of the serial port on occasions, especially when we boot up? There is random out of band data sent to the serial uart for debugging and monitoring purposes.

AHHams commented:
@wolfmanjm: thanks for the pointers. Ill hopefully get around to working on it next weekend. 

I asked David about zprobe: it just needs to send the state of the probe. Triggered / not triggered. Any suggestions how to pick that up?

Indeed it uses the serial uart. PanelDues behaviour during bootup, waiting for Smooth is indeed random. I sometimes have to reset the PanelDue after Smoothie bootup, after which it is stable. Sometimes it seems to work on the first go.

wolfmanjm commented:
You need to look at the zprobescreen in the panel code to see how to get the zprobe status.
PanelDue needs to ignore anything it receives that is not properly formed Json as the uart is used to send log messages.

wolfmanjm commented:
I think you can get Fan PWM using `PublicData::get_value( switch_checksum, get_checksum("fan"), value_checksum, &v ); ` If the user uses a convention to call the `switch.fan.enable` in the config. If not there is no way to figure out what is a fan and what is not. Also PWM is not RPM, there is no way to map PWM to RPM so it will need to be percent fan speed where PWM of 255 would be 100%

AHHams commented:
@wolfmanjm the zprobescreen in the panel just issues a m119 and displays the returned string. I can use that approach, but that would require me to parse the returned string. Which seems wasteful both cpu and memory wise. Is there something against just declaring a pin in the same way as in zProbe.cpp and get the state that way?

wolfmanjm commented:
No. You cannot declare a pin twice. Issuing M119 and displaying the message is the best approach. Or create a public message to get the probe status.

AHHams committed:
Merge remote-tracking branch 'Smoothieware/edge' into feature/due_panel

AHHams committed:
implement probe output on M408

Code works. Tested with and without probe enabled. Code is not optimal.

AHHams committed:
Merge remote-tracking branch 'Smoothieware/edge' into feature/due_panel

AHHams committed:
Code style fixes. No code change.

Run the source through uncrustify to get style consistent with rest of
Smoothieware source.

AHHams commented:
@wolfmanjm: I figured out why the fan output I made using PublicData call always returned 100% when fan is on. Apparently the pad_switch (from Switch's public data) struct's "value" is the max_pwm rather than the current pwm_pin value. Appears by design, not bug, of Switch.cpp.

Probe output now works. With sub-optimal code, parsing M119 output.

AHHams commented:
@wolfmanjm: The Reporter for M408 is already quite functional as it is now. To make it work fully, requires changes outside the scope of the Reporter.cpp file.

To make the SD card file display working, either:
1. Smoothie needs to implement:
- [M20 S2 ](http://reprap.org/wiki/G-code#M20:_List_SD_card)
- [M36](http://reprap.org/wiki/G-code#M36:_Return_file_information), or
2. PanelDue firmware is to work with normal M20 output. 

I think there is value to these JSON outputs, given they may be used also by web interfaces. Then again, rather than making all kind of different modules output JSON, it may be a nicer design if there is one module? What do you is the best approach?

wolfmanjm commented:
both M36 and M20 S2 are illegal gcodes. you cannot pass filenames and such as gcode. We are trying to kkeep the gcode parser to be a standard gcode and no exceptions (there are currently only two exceptions). The lowercase commands (like ls /sd) are designed for this kind of thing.
My feeling is paneldue should use the file list that has been defined for years and not try to recreate new codes, and ones that are illegal.

AHHams committed:
implement "numTools"

Implemented numTools.

AHHams committed:
Let Reporter handle JSON output option for M20

In future version of PanelDue, JSON out for M20 will go to “M408 S20”,
but for now, let’s simply make this ignore “M20 S2” option, and let
Reporter code handle it.

AHHams committed:
Merge remote-tracking branch 'Smoothieware/edge' into feature/due_panel

AHHams commented:
@wolfmanjm I agree that it is not good design to simply start adding JSON output to various existing GCODES. Then again, I think adding JSON output is a good idea as a allows more structured/robust form of communication with web interfaces, etc. (rather the parsing existing outputs. e.g., of "ls"). 

I think is it is best to use M408 for all JSON output and control it through switches on M408. 

Chris is working on a new version of the PanelDue firmware that will call for the JSON output in this way. See discussion here: [PanelDue issue 5](https://github.com/dc42/PanelDue/issues/5).

For now (and please see this as a temporary hack), I do implement M36 and M20 S2 in this pull request to allow testing of the dir listing / file info.

AHHams commented:
I have not been able to get one key feature of the PanelDue to work yet: the console. 

To support this, I need to support: 
>seq:     the sequence number of the most recent non-trivial G-code response or error message. Only present if the R parameter was provided and the current sequence number is greater than that.
> resp:    the most recent non-trivial G-code response or error message. Only present if the R parameter was provided and the current sequence number is greater.

Any suggestion for how to go about implementing this?

arthurwolf commented:
Hey !

At this point I believe enough paneldue features are implemented that this can be generally useful to paneldue users, even with some features missing.
I propose we merge this into mainstream, advertise it so this gets more users, and from there, we can work on building better suppport.

What do you think ? This has been hanging in the air for a very long time, and I get frequent requests for this to be merged.

Cheers.

AHHams commented:
Sounds good. Let me make some final checks this weekend.

AHHams committed:
Merge remote-tracking branch 'Smoothieware/edge' into feature/due_panel

dc42 commented:
The latest release of PanelDue firmware 1.16beta1 here https://github.com/dc42/PanelDueFirmware/tree/master/Release/Edge includes code to modify the behaviour when the host announces its firmware type as "Smoothie" or anything that begins with that. The changes are:
- Suppress the standby temperature fields
- Use M104 instead of G10 to send the active temperatures
- Strip the drive letter from the path before sending paths and filenames
- Use M408 S20 instead of M20 S2, and M408 S36 instead of M36
This is not tested with Smoothieware and probably contains bugs.

AHHams committed:
Merge remote-tracking branch 'Smoothieware/edge' into feature/due_panel

AHHams commented:
I just checked 1.16beta1 of the PanelDue firmware and brought the pull request back in sync with edge. Everything appears to be work fine. 

@dc42: I have some observations on 1.16beta1:
- Setting hot end temperature now works.
- File browser sometimes works / sometimes does not. Opening / closing window appears to make it work / not work. 
- "Files on card" always shows 0, even if it returns a list of files.
- Standby temps is still there (for bed).
- "Move" window sometimes closes when moving and PanelDue then automatically opens the "Print" screen.

In general, things that _do not work_ yet (because of current state of this M408 implementation):
- Console
- File information
- Deleting a file

dc42 commented:
I'm glad it's working better for you. The number after "Files on card" is the number of the SD card, not the number of files, because Panel Due supports more than one SD card. I can fix the bed standby temperature appearing in the 1.16 release which is very close. Panel Due automatically switches to the Print window when the status changes to Printing, so please check that Smoothieware is sending the right codes. The status should only be returned as Printing when printing a file from the SD card. Regarding the file browser, please check that all the M408 responses you send end in newline. HTH David

AHHams commented:
@dc42: I just noted that I had not switched the code over to using M408 S20 and M408 S36. I.e., it is still using M20 and M36. Yet file browsing still works. When I switch over the code to M408 S20 and M408 S36, file browsing no longer works. I guess PanelDue is still using M20 and M36 for Smoothie. I think there is a bug "Filemanager.cpp", line 49. There should be a: ((GetFirmwareFeatures() & noM20M36) != 0) ? "M408 S20 P" : "M20 S2 P")

dc42 commented:
The command is changed to M408 at line 213 when the host is Smoothieware so the fault must lie elsewhere. I'll look into it.

dc42 commented:
@AHHams: Please try the 1.16beta2 firmware now at https://github.com/dc42/PanelDueFirmware/tree/master/Release/Edge.

AHHams committed:
Merge remote-tracking branch 'Smoothieware/edge' into feature/due_panel

AHHams committed:
Remove handling of M20 and M36

Removing handling of M20 and M36, because PanelDue 1.16beta2 supports
M408 S20 and M408 S36 respectively.

AHHams commented:
@dc42: I checked out 1.16beta2 and file browsing now works fine. 

I committed the change to the reporter that will implement M408 S20 and M408 S36. It no longer implement M20 and M36 codes. This means that this pull request is now entirely local. No changes to existing Smoothieware code is needed.

Only one small thing to do is roll back change I made to SimpleShell.cpp. I was looking how to do that the "git way", by simply removing my original commit of the change, but I did not yet figure out how to do that. Next time I have to look at it is next weekend.

triffid commented:
git revert ;)

On 3 April 2017 at 05:34, AHHams <notifications@github.com> wrote:

> @dc42 <https://github.com/dc42>: I checked out 1.16beta2 and file
> browsing now works fine.
>
> I committed the change to the reporter that will implement M408 S20 and
> M408 S36. It no longer implement M20 and M36 codes. This means that this
> pull request is now entirely local. No changes to existing Smoothieware
> code is needed.
>
> Only one small thing to do is roll back change I made to SimpleShell.cpp.
> I was looking how to do that the "git way", by simply removing my original
> commit of the change, but I did not yet figure out how to do that. Next
> time I have to look at it is next weekend.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Smoothieware/Smoothieware/pull/1059#issuecomment-291016893>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAKBGgFqoSzGi3XuHrzG5O8_IEGjtXXbks5rsBRYgaJpZM4Ks-zq>
> .
>

AHHams committed:
Merge remote-tracking branch 'Smoothieware/edge' into feature/due_panel

AHHams committed:
Reverting SImpleShell.cpp

Undoing change made to SimpleShell.cpp. Now same again as in edge.

AHHams commented:
@triffid: Thanks. The original commit made changes to multiple files. I did not find how to use revert for change to one file only. So I used `git checkout Smoothieware/edge src/modules/utils/simpleshell/SimpleShell.cpp`. That works.

AHHams commented:
@wolfmanjm: I am currently using `!THEKERNEL->conveyor->is_queue_empty()` to assess whether Smoothie is "playing". However, when jogging repeatedly it can then happen that the queue is not empty when this code runs, so the Reporter mistakes jogging for a file being played from the SD card. Is there a better way to check a file being played?

wolfmanjm commented:
that is deprecated and not correct use is_idle() instead.

AHHams committed:
Use is_idle() instead of deprecated is_queue_empty()

Switching to is_idle().

AHHams commented:
@wolfmanjm: I switched to `is_idle()`, but experience the same problem: when jogging repeatedly, is_idle() returns false, making the reporting return the "Playing" state. Is there a different way of establishing that a file is being played from SD Card?

AHHams commented:
@dc42: You wrote above "The status should only be returned as Printing when printing a file from the SD card.". It looks like I can only return 'P' when Smoothie is not idle. I have no way of determining whether Smoothie is playing something from SD card, or simply still busy executing jogging (or other commands) done from the PanelDue. Is it possible to change the rule on PanelDue's end (if need be, for Smoothie only)? E.g., only automatically switch to Playing screen when PanelDue receives 'P' _and_ no user input has been received for X (say 3) seconds?

wolfmanjm commented:
@AHHams  look at the watchscreen.cpp code to see how to determine if printing from sdcard or not. on_idle() won't do that, and is not supposed to do that .

AHHams committed:
Fixed Playing bug

Fixed bug where Reporter would indicate playing status when merely
busy. Used code from Panel.

AHHams commented:
@wolfmanjm: Thanks. Problem fixed. I copied code from Panel.ccp (`is_playing()`).
@dc42: Please ignore my last suggestion.

AHHams commented:
There is a problem with playing files from SD Card. When using "M32" to print (as PanelDue does), Smoothie prepends "/sd/" in front of the path (causing file not to be found. E.g., it tries to print "/sd//sd/home.gco"). Smoothie also recognises "play" (also used by Panel, FileScreen.cpp) which does not prepend "/sd/". 

I see two solutions:
1. I change Reporter code so "/sd/" becomes the root, or
2. PanelDue uses "play" to play files when it sees Smoothieware. 

I would have preference for latter as it allows browsing and playing files outside /sd. 

Any views?

wolfmanjm commented:
FWIW files are NEVER outside of /sd. And if you have an external sdcard you should NEVER try to play files from it.

AHHams commented:
Thanks. That makes it easy then: I'll make "/sd" the root for browsing files (option 1).

AHHams committed:
Merge remote-tracking branch 'Smoothieware/edge' into feature/due_panel

arthurwolf commented:
A few users have expressed interrest in testing this and reporting here, would you mind making a .bin and posting it somewhere users can easily download it and posting the link here ?
Thanks ! :)

AHHams committed:
Merge remote-tracking branch 'Smoothieware/edge' into feature/due_panel

AHHams committed:
Partial fix for make M408 S20 and M408 S36 behave as if /sd is root

There is still a bug somewhere related to this fix.

AHHams commented:
@arthurwolf: Here it is: [SmoothiePanelDue.bin](https://www.dropbox.com/s/igohr9f4oil04wu/SmoothiePanelDue.bin?dl=0)

Note that I am still working on fully ironing out a bug with the path handling of M408 S20 and M408 S36. Browsing Smoothie's SD CARD and Printing from it should work, but it is a bit flaky...

arthurwolf commented:
Thanks a ton.

dc42 commented:
Anthony, let me know if you want me to include the /sd in the file path sent by PanelDue. It already uses /gcodes as the root when talking to RRF, so I can easily make it use /sd as the root when talking to Smoothieware.

AHHams commented:
@dc42: Thanks for the offer. But since Smoothie already prepends `"/sd"` automatically to the path passed to `M32`, it seems more logical to me that the behaviour of `M408 S20` and `M408 S36` is made consistent with that. 

Questions for you though:
1) When you send a path from PanelDue (to Smoothie), does it always start with a '/'?
2) When the path is a directory, does it always end in '/' or just the name of the directory?

(I am asking because some of the inconsistency I am seeing may also relate to what PanelDue sends as the path string. Would be good if I could rule that out).

AHHams committed:
Merge remote-tracking branch 'Smoothieware/edge' into feature/due_panel

dc42 commented:
When the host is Smoothieware, the root path is set to "0:/" instead of "0:/gcodes/", then for Smoothieware the "0:" is also stripped from the beginning of the path when sending the command. So paths sent to Smothieware should always start with "/". Let me know if you think it is doing something different.

AHHams commented:
The first browse of the root works fine (with my latest commit I always prepend '/sd' - without a finishing '/'). When I click on a file in PanelDue, it appears as if PanelDue sent (after prepending '/sd') '/sd//home.gco'. I.e., I get a doubling of a slash. Probably a bug in the latest my latest commit. I am just having trouble finding where I went wrong.

dc42 commented:
@AHHams: What are you sending in the "dir" variable in the M408 S20 response? That may have a bearing on the issue, because PanelDue uses it to build the full path of the file you select to print.

AHHams commented:
Thanks David. I'll focus on that first (in two weeks, when I have time to look at it again).

Zedsquared commented:
Hi Folks,  I now have a PanelDue to go with my smoothieboard driven printer and want to help test this. I have the latest firmware for the panel from https://github.com/dc42/PanelDueFirmware/tree/master/Release/Stable and the smoothie build linked above. What config needs changing? I assume enabling the uart and setting the baud rate, what else?

dc42 commented:
Zedsquared, you'll need to upgrade the PanelDue firmware to version 1.16 if it isn't already running that version. AHHams can best advise you on the Smoothie side.

AHHams commented:
This pull request is not in latest edge yet. If you want to give it a try (at your own risk, etc) use the firmware I compiled and linked to a few comments above.  For settings, look at the config example file in this pull request.

Zedsquared commented:
So, correct me if I'm wrong, config just needs this adding: 

reporter.enable                 true
reporter.machine_name    myMachine
uart0.baud_rate                 57600 

Easier than I expected :)

Zedsquared commented:
OK so I have hooked up the panel (with v1.16 paneldue firmware and smoothieware linked to above)  and can file browse, jog, home etc nicely. 

File browsing doesn't always work first time, closing the empty file list then browsing again seems to work. 

Clicking on a file will show file name plus (I presume) placeholder info for height, layer height etc. The "sliced by" field shows the file path with a double slash after /sd as @AHHams mentioned above. Hitting print seems to start the file printing ok but there is no update of temperatures (either live or set) etc on the "print" screen. Also the printer status shows as 'idle' while it is printing the selected file.

Good start!

paboman commented:
I have a smoothieboard and the paneldue 3.0 with 4.3 LCD, are there any instructions on how to connect them together ?
thx

wolfmanjm commented:
This appears to be a dead PR.
